File: 2015/js/2015-05.js
const path  = require('path')
const fs    = require('fs')

const data = fs.readFileSync(path.resolve(__dirname, '../data/2015-05.txt'), 'utf8')

function isNice1(input) {
  const vowels    = ['a', 'e', 'i', 'o', 'u']
  const blacklist = ['ab', 'cd', 'pq', 'xy']

  const vowelCount     = input.split('').filter(c => vowels.includes(c)).length
  const duplicateCount = input.split('').filter((c, i) => input[i + 1] == c).length
  const blacklistCount = blacklist.filter(w => input.includes(w)).length

  return vowelCount > 2 && duplicateCount > 0 && blacklistCount == 0
}

function isNice2(input) {
  let duplicateCount = 0

  input.split('').forEach((c, i) => {
    const restOfWord = input.substring(i+2, input.length)
    const chars = input[i]+input[i+1]
    if (restOfWord.includes(chars)) duplicateCount++
  })

  const spaceBetweenCount = input.split('').filter((c, i) => input[i + 2] == c).length

  return  duplicateCount > 0 && spaceBetweenCount > 0
}

console.clear()

console.assert( isNice1('ugknbfddgicrmopn') == true,  'Part 1 Test 1' )
console.assert( isNice1('aaa')              == true,  'Part 1 Test 2' )
console.assert( isNice1('jchzalrnumimnmhp') == false, 'Part 1 Test 3' )
console.assert( isNice1('haegwjzuvuyypxyu') == false, 'Part 1 Test 4' )
console.assert( isNice1('dvszwmarrgswjxmb') == false, 'Part 1 Test 5' )

console.assert( isNice2('qjhvhtzxzqqjkmpb') == true,  'Part 2 Test 1' )
console.assert( isNice2('xxyxx')            == true,  'Part 2 Test 2' )
console.assert( isNice2('uurcxstgmygtbstg') == false, 'Part 2 Test 3' )
console.assert( isNice2('ieodomkazucvgmuy') == false, 'Part 2 Test 4' )

console.log( `
Advent of code 2015 #05
Part1: ${data.split('\n').filter(word => isNice1(word)).length}
Part2: ${data.split('\n').filter(word => isNice2(word)).length}
` )

File: 2015/c/2015-05.c
#include <stdio.h>
#include <string.h>

#include "tools.h"

int part1(char input[]) {
  char data[strlen(input)];
  const char vowels[5] = {'a', 'e', 'i', 'o', 'u'};
  const char blackList[4][2] = {"ab", "cd", "pq", "xy"};
  int niceCount = 0;

  strcpy(data, input);
  char *line = strtok(data, "\n");

  while (line != NULL) {
    int vowelCount     = 0;
    int doublesCount   = 0;
    int blackListCount = 0;

    for (int i = 0; i < strlen(line); i++) {
      // Find vowels
      for (int j = 0; j < sizeof(vowels); j++)
        if (line[i] == vowels[j]) vowelCount++;
    
      if (i+1 < strlen(line)) {
        // Find doubles
        if (line[i] == line[i+1]) doublesCount++;
    
        // Find blacklisted
        for (int j = 0; j < 4; j++) {
          if (line[i] == blackList[j][0] && line[i+1] == blackList[j][1]) blackListCount++;
        }

      }
    }

    if (vowelCount > 2 && doublesCount > 0 && blackListCount == 0) niceCount++;
    
    line = strtok(NULL, "\n");
  }
  
  return niceCount;
}

int part2(char input[]) {
  char data[strlen(input)];
  int niceCount = 0;

  strcpy(data, input);
  char *line = strtok(data, "\n");

  while (line != NULL) {
    int duplicationCount  = 0;
    int spaceBetweenCount = 0;

    for (int i = 0; i < strlen(line); i++) {
      char chars[2] = {line[i], line[i+1]};

      for (int j = i+2; j < strlen(line); j++) {
        if (line[j] == chars[0] && line[j+1] == chars[1]) duplicationCount++;
      }
      
      if (i+2 < strlen(line) && line[i] == line[i+2]) spaceBetweenCount++;
    }

    if (duplicationCount && spaceBetweenCount) niceCount++;

    line = strtok(NULL, "\n");
  }


  return niceCount;
}

int main() {
  char *data = getDataFromFile("../data/2015-05.txt");

  printf("Advent of code 2015 #5\n\n");

  printf("----Part 1----\n");
  test("Test 1", part1("ugknbfddgicrmopn"), 1);
  test("Test 2", part1("aaa"),              1);
  test("Test 3", part1("jchzalrnumimnmhp"), 0);
  test("Test 4", part1("haegwjzuvuyypxyu"), 0);
  test("Test 5", part1("dvszwmarrgswjxmb"), 0);
  printf("Answer: %d\n\n", part1(data));

  printf("----Part 2----\n");
  test("Test 1", part2("qjhvhtzxzqqjkmpb"), 1);
  test("Test 2", part2("xxyxx"),            1);
  test("Test 3", part2("uurcxstgmygtbstg"), 0);
  test("Test 4", part2("ieodomkazucvgmuy"), 0);
  printf("Answer: %d\n\n", part2(data));

  return 0;
}